app/posthandler/refund_gas_remaining.go

```diff
@@ -0,0 +1,155 @@
+	if err := d.maybeRefund(ctx, tx, simulate); err != nil {
+		return ctx, err
+	}
+	return next(ctx, tx, simulate)
+}
+
+// maybeRefund conditionally refunds a portion of the tx fee to the fee payer.
+func (d RefundGasRemainingDecorator) maybeRefund(ctx sdk.Context, tx sdk.Tx, simulate bool) error {
+	// If this tx was included in a v1 block then no refund needs to be issued.
+	// Gas refunds are only applicable for blocks >= v2.Version
+	if ctx.BlockHeader().Version.App == v1.Version {
+		return nil
+	}
+
+	// If this is a simulation then no state changes should be persisted and
+	// therefore no refund needs to be issued.
+	if simulate {
+		return nil
+	}
+
+	// Replace the context's gas meter with an infinite gas meter so that this
+	// posthandler doesn't run out of gas while refunding.
+	gasMeter := ctx.GasMeter()
+	ctx = ctx.WithGasMeter(sdk.NewInfiniteGasMeter())
+	// Restore the original gas meter after this posthandler is done.
+	defer ctx.WithGasMeter(gasMeter)
+
+	// If the gas meter doesn't have enough gas remaining to cover the
+	// refund gas cost, then no refund needs to be issued.
+	if gasMeter.GasRemaining() < RefundGasCost {
+		return nil
+	}
+	gasMeter.ConsumeGas(RefundGasCost, "refund gas cost")
+
+	feeTx, ok := tx.(sdk.FeeTx)
+	if !ok {
+		return errors.Wrap(sdkerrors.ErrTxDecode, "tx must be a FeeTx to use FeeRefundDecorator")
+	}
+
+	refund := getRefund(gasMeter, feeTx)
+	if len(refund) == 0 {
+		return nil
+	}
+	recipient := getRecipient(feeTx)
+
+	if err := d.processRefund(ctx, refund, recipient); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// processRefund sends amountToRefund from the fee collector module account to the recipient.
+func (d RefundGasRemainingDecorator) processRefund(ctx sdk.Context, refund sdk.Coins, recipient sdk.AccAddress) error {
+	from := d.accountKeeper.GetModuleAddress(authtypes.FeeCollectorName)
+	if from == nil {
+		return fmt.Errorf("fee collector module account (%s) has not been set", authtypes.FeeCollectorName)
+	}
+
+	if recipientAccount := d.accountKeeper.GetAccount(ctx, recipient); recipientAccount == nil {
+		return errors.Wrapf(sdkerrors.ErrUnknownAddress, "recipient address: %s does not exist", recipientAccount)
+	}
+
+	if !refund.IsValid() {
+		return fmt.Errorf("invalid coins to refund: %s", refund)
+	}
+
+	if err := d.bankKeeper.SendCoins(ctx, from, recipient, refund); err != nil {
+		return errors.Wrapf(err, "error refunding %s from fee collector module account to %s", refund, recipient)
+	}
+
+	return nil
+}
+
+// getRefund returns the coins that should be refunded to the recipient.
+func getRefund(gasMeter sdk.GasMeter, feeTx sdk.FeeTx) sdk.Coins {
+	gasPrice := getGasPrice(feeTx)
+	toRefund := gasPrice.Amount.MulInt(sdk.NewIntFromUint64(gasMeter.GasRemaining())).TruncateInt()
+	maxToRefund := MaxPortionOfFeeToRefund.MulInt(feeTx.GetFee().AmountOf(appconsts.BondDenom)).TruncateInt()
```

Comment:
[block]

So I think this should be proportional to the gas used * gas price not the fee. If someone submitted a send tx with a gas limit of 50 million (or whatever is needed to full a block) they would prevent others from using the block at half the gas price they provided. If we use `gas_used * gas_price` then they will get charged basically the full price as they will only get refunded for the 100,000 + 50% of the gas used

Reply:
Sorry I don't quite follow your example. Where does 100,000 come from? 

> So I think this should be proportional to the gas used * gas price not the fee.

this refers to `maxToRefund`? If yes are you proposing something like:

```go
maxToRefund := gas_used * gas_price * SomeConstant
```

Reply:
100,00 in this example is the actual gas used for the message

Reply:
Yes I'm proposing something like this

Reply:
Got it. Here are some names + numbers we can use to make it clear what we're referring to

```go
// constants
MaxPortionOfFee := .5
MaxPortionOfFeeUtilizedToRefund := .5

// user specified
gasLimit := 50,000,000
gasPrice := 1
fees := 50,000,000
// execution of tx
gasUsed := 100,000

// existing method
maxToRefund := MaxPortionOfFee * fees // 25,000,000
// proposed method
maxToRefund := gasUsed * gasPrice * MaxPortionOfFeeUtilizedToRefund // 50,000
```

Reply:
Good point. This is actually the proposed design in https://github.com/cosmos/cosmos-sdk/issues/2150#issue-354029283 and setting `MaxPortionOfFeeUtilizedToRefund` to 1.5 or 2

Reply:
Yes exactly!

Reply:
Resolved in 161664b11a22b089c2ba823691b18dc15b1456e0 but need to improve tests.